### ISSUE TITLE
possible rsa-sha2-512 signature issue in 0.75+?

### ISSUE THREAD

**Issue by MetricMike** [NoMoreFood/putty-cac#77]
When connecting from a Win10 client to a CentOS7 (OpenSSH 7.4) server, I'm seeing the following error get thrown:

(from client):
```
Authenticating with public key "CAPI:...."
Server refused public-key signature despite accepting key!
(pop up - No Supported authentication methods available (server sent: publickey,gssapi-keyex,...))
```

(from server):
```
debug1: matching key found file ..../.ssh/authorized_keys, line 4 RSA SHA256:....
...
debug3: mm_answer_keyallowed: key ... is allowed
debug3: mm_request_send entering: type 23
debug3: mm_key_verify entering [preauth]
debug3: mm_request_send entering: type 24 [preauth]
debug3: mm_key_verify: waiting for MONITOR_ANS_KEYVERIFY [preauth]
debug3: mm_request_receive_expect entering: type 25 [preauth]
debug3: mm_request_receive entering [preauth]
debug3: mm_request_receive entering
debug3: monitor_read: checking request 24
error: key_verify: invalid argument
debug3: mm_answer_keyverify: key ... signature unverified
debug3: mm_request_send: entering type 25
Failed publickey for ... from ... port ... ssh2: RSA SHA256: ...
debug2: userauth_pubkey: authenticated 0 pkalg rsa-sha2-512 [preauth]
userauth_finish: failure partial=0 next methods="publickey,gssappi..." [preauth]
```

Loading CAPI certs in pageant-cac or directly in putty-cac exhibits the same behavior. Reverting to putty-cac 0.74 works, but I note that the algorithm is ssh-rsa there. I believe the change in putty 0.75 which transparently defaults to SHA-256/512 if both client and server declare support (https://www.chiark.greenend.org.uk/~sgtatham/putty/wishlist/rsa-sha2.html) is what exposed this issue.

This might be related to the initial support introduced in putty-cac 0.74 on https://github.com/NoMoreFood/putty-cac/pull/56 which indicated that there may be additional work needed but everything looked fine at the time.

Using a "traditionally" generated RSA 4096-bit key works great (I see `mm_answer_keyverify: key ... signature verified` and `userauth_pubkey: authenticated 1 pkalg rsa-sha2-512 [preauth]` messages), so I think it's the specific code introduced for CAPI(/PKCS too?) support that's causing the issue. 

Does that sound plausible / are there any workarounds / any additional information I could provide to help resolve this?

**Comment by NoMoreFood** [comment]
Do you see the same behavior with 2048-bit key?  Can you produce the issue with a self-signed software-backed certificate?

**Comment by MetricMike** [comment]
Self-signed CAPI certificates do not appear to have this problem after cloning them from the "official" certs I'm trying to use.

I should note the official certs are issued/signed by an internal corporate CA but that CA is also in the server's default trust bundle.

**Comment by NoMoreFood** [comment]
Did you install the self-signed certs onto the hardware token?  If not, is it possible the hardware token you using (or the Windows mini-driver used to interact with it) does not support SHA2 signatures?

**Comment by MetricMike** [comment]
These are software-only certs, so no hardware token, loaded directly into the default Windows store.

The nonworking certificates are SHA384, but a self-signed SHA384 with the exact same attributes loaded into the same store works just fine. The `key_verify: invalid argument` feels like the only pointer I have to what's actually going wrong.

**Comment by NoMoreFood** [comment]
Interesting.  The cert chain should make zero difference in terms of the signing operations.  Can you provide a certutil dump of a the working / non-working certificate?

**Comment by MetricMike** [comment]
I can try! Might be a few days to get a reproducible sample I can post unredacted, but will ping here when I do.

**Comment by MetricMike** [comment]
I'm still having trouble reproducing outside of one specific client/server setup that caused the initial report, but further googling on that line leads me to a bug (https://bugzilla.mindrot.org/show_bug.cgi?id=2799) in the signature key exchange that was fixed in OpenSSH 7.8 (which won't be backported to RHEL/CentOS7 distros because it's not a security fix).

So, I think the SHA2 sig is failing for some reason (likely something proprietary that I have in this setup), and the fallback mechanism relies on the bz#2799 available on the server.

If that's the case, I don't think it's reasonable to ask putty-cac to support a 5-year old version of openssh and this can probably get closed.

**Comment by NoMoreFood** [comment]
Sounds good.  Let me know if you see it in newer releases.
